src/metabase/db/custom_migrations/metrics_v2.clj

```diff
@@ -0,0 +1,158 @@
+      (cond-> query
+        (not= rewritten aggregation) (assoc :aggregation rewritten)))))
+
+(defn- rewrite-metric-consuming-card
+  "Rewrite the question `card` replacing references to v1 metrics with references
+  to the corresponding v2 metric question as specified by the mapping `metric-id->metric-card`."
+  [card metric-id->metric-card]
+  (let [dataset-query (json/parse-string (:dataset_query card) true)
+        inner-query (:query dataset-query)
+        rewritten (rewrite-metric-consuming-query inner-query metric-id->metric-card)]
+    (when (not= rewritten inner-query)
+      (-> card
+          (assoc :dataset_query (-> dataset-query
+                                    (assoc :query rewritten)
+                                    json/generate-string))
+          (dissoc :updated_at)))))
+
+(defn migrate-up
+  "Migrate metrics and the cards consuming them to metrics v2. This involves
+  1. creating the migration collection,
+  2. creating a v2 metric card for each v1 metric in the migration collection,
+  3. making a backup copy of the dataset_query of each metric consuming card,
+  4. rewriting the metric consuming cards to use the migrated v2 metric cards."
+  []
+  (when (t2/exists? :metric)
+    (let [metric-v2-coll (get-metric-migration-collection {:create? true})
+          metric-id->metric-card (into {}
+                                       (map (juxt :id #(create-metric-v2 % metric-v2-coll)))
+                                       (t2/query {:select [:m.* [:t.db_id :database_id]]
+                                                  :from [[:metric :m]]
+                                                  :left-join [[:metabase_table :t] [:= :t.id :m.table_id]]}))
+          metric-consuming-cards (t2/select :report_card {:where [:like [:lower :dataset_query] "%[\"metric\",%"]})]
+      (doseq [card metric-consuming-cards
+              :let [rewritten (rewrite-metric-consuming-card card metric-id->metric-card)]
+              :when (and rewritten (not= card rewritten))]
+        (t2/update! :report_card
+                    (:id card)
+                    (-> rewritten
+                        (assoc :dataset_query_metrics_v2_migration_backup (:dataset_query card)
+                               :updated_at (Instant/now))))))))
+
+(defn- try-deleting-collection
+  [coll]
+  (try
+    (t2/delete! :collection :id (:id coll))
+    (catch Exception _
+      (log/warn "couldn't delete migration collection" (:name coll)))))
+
+(defn migrate-down
+  "Revert the migration to v2 metrics. This involves
+  1. restoring the metric consuming cards from the backup,
+  2. deleting the metric cards created by the up migration,
+  3. deleting the metric migration collection."
+  []
+  (when-let [metric-v2-coll (get-metric-migration-collection)]
+    (doseq [{:keys [id dataset_query_metrics_v2_migration_backup]}
+            (t2/select :report_card {:where [:!= :dataset_query_metrics_v2_migration_backup nil]})]
```

Comment:
I was going to say use `t2/reducible-select` + `reduce` here, but this entire migration can be done in SQL, can't it? 

This is super inefficient, selecting the entire row for every card that has a non-`NULL` `dataset_query_metrics_v2_migration_backup` in memory all at once, and then one by one updating their value of `dataset_query`... right? 

You can just do 

```sql
UPDATE report_card
SET dataset_query = dataset_query_metrics_v2_migration_backup
WHERE dataset_query_metrics_v2_migration_backup IS NOT NULL
```

That's one query, instead of potentially thousands, and it can all be done in SQL (Liquibase) instead of in Clojure-land.

Reply:
Agree, this is also a residual thing.